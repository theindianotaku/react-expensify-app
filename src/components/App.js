import React, { Component } from 'react';


class App extends Component {
  render() {
    return (
      <div>
        <p>Hello there! Expensify here.</p>
      </div>
    );
  }
}

export default App;
