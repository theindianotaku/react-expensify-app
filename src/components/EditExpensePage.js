import React from 'react';

const EditExpensePage = (props) => (
  <div>
    <p>Hello! Edit page of id {props.match.params.id} here here.</p>
  </div>
);

export default EditExpensePage;
