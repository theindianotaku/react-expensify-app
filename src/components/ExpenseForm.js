import React, { Component } from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends Component {
  constructor (props) {
    super(props);
    this.state = {
      amount: props.expense ? (props.expense.amount/100) : '',
      createdAt: props.expense? moment(props.expense.createdAt) : moment(),
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      focused: false,
      error: ''
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({
      description
    }));
  }

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({note}));
  }

  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  }

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  }

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ focused }));
  }

  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'Please provide description and amount.'
      }));
    } else {
      this.setState(() => ({error: ''}));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <form
        action=""
        className="form"
        onSubmit={this.onSubmit}
      >
        {this.state.error && <p className="form__error">{this.state.error}</p>}

        <input
          type="text"
          placeholder="Description"
          className="text-input"
          autoFocus
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <input
          type="text"
          placeholder="Amount"
          className="text-input"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />

        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.focused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false }
        />

        <textarea
          className="textarea"
          placeholder="Add a note for your expense (optional)"
          value={this.state.note}
          onChange={this.onNoteChange}
        >
        </textarea>
        <button className="button dark" type="Submit">Save Expense</button>
      </form>
    );
  }
}

export default ExpenseForm;
