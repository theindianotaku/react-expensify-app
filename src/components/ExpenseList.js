import React from 'react';

const ExpenseList = () => (
  <div>
    Expense List
  </div>
);

export default ExpenseList;
