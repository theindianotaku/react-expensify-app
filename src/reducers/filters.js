const defaultState = {
  text: '',
  sortBy: 'date',
  startDate: undefined,
  endDate: undefined
};

// state = state.filters
const filtersReducer = (state = defaultState, action) => {
  switch (action.type) {
    case 'SET_TEXT_FILTER':
      return {
        ...state,
        text: action.text
      };
    case 'SET_START_DATE':
      return {
        ...state,
        startDate: action.startDate
      };
    case 'SET_END_DATE':
      return {
        ...state,
        endDate: action.endDate
      };
    case 'SORT_BY_DATE':
      return {
        ...state,
        sortBy: 'date'
      };
    case 'SORT_BY_AMOUNT':
      return {
        ...state,
        sortBy: 'amount'
      };
    default:
      return state;
  }
};

export default filtersReducer;
