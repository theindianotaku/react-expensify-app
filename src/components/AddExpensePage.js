import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import { startAddExpense } from '../thunks/expenses';
import ExpenseForm from './ExpenseForm';

export class AddExpensePage extends Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <p className="page-header__title">Add Expense</p>
            <div className="page-header__actions">
              <Link className="button simple valign-wrapper" to="/dashboard">Go back</Link>
            </div>
          </div>
        </div>
        <div className="content-container m-b-l">
          <ExpenseForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
