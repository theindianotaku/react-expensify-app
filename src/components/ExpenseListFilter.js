import React, { Component } from 'react';
import { connect } from 'react-redux';
import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilter extends Component {
  state = {
    calendarFocussed: null
  };

  onDatesChange = ({startDate, endDate}) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (calendarFocussed) => {
    this.setState(() => ({ calendarFocussed }));
  };

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };

  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };

  render () {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              type="text"
              placeholder="Search expenses"
              className="text-input"
              value={this.props.filters.text}
              onChange={this.onTextChange}
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              onChange={this.onSortChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate = {this.props.filters.startDate}
              endDate = {this.props.filters.endDate}
              onDatesChange = {this.onDatesChange}
              focusedInput = {this.state.calendarFocussed}
              onFocusChange = {this.onFocusChange}
              showClearDates = {true}
              numberOfMonths = {1}
              isOutsideRange={() => false}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  };
};

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);
