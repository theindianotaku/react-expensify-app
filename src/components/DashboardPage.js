import React, { Component } from 'react';

import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilter';
import ExpenseSummary from './ExpenseSummary';

class DashboardPage extends Component {
  render() {
    return (
      <div>
        <ExpenseSummary />
        <ExpenseListFilter />
        <ExpenseList />
      </div>
    );
  }
}

export default DashboardPage;
