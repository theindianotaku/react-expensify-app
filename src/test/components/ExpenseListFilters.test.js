import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseListFilter } from '../../components/ExpenseListFilter';
import { filters, altFilters } from '../fixtures/filters';

let setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate, wrapper ;

beforeEach(() => {
  setTextFilter = jest.fn();
  sortByDate = jest.fn();
  sortByAmount = jest.fn();
  setStartDate = jest.fn();
  setEndDate = jest.fn();
  wrapper = shallow(
    <ExpenseListFilter
      setTextFilter={setTextFilter}
      sortByDate={sortByDate}
      sortByAmount={sortByAmount}
      setStartDate={setStartDate}
      setEndDate={setEndDate}
      filters={filters}
    />
  );
});

test('should render ExpenseListFilter properly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseListFilter with alt data properly', () => {
  wrapper.setProps({
    filters: altFilters
  });
  expect(wrapper).toMatchSnapshot();
});

test('should handle text change', () => {
  const value = 'test';
  wrapper.find('input').simulate('change', {
    target: { value }
  });
  expect(setTextFilter).toHaveBeenLastCalledWith(value);
});

test('should set sort by date', () => {
  const value = 'date';
  const e = {
    target: { value }
  };
  wrapper.find('select').simulate('change', e);
  expect(sortByDate).toHaveBeenCalled();
});

test('should set sort by date', () => {
  const value = 'amount';
  const e = {
    target: { value }
  };
  wrapper.setProps({
    filters: altFilters
  });
  wrapper.find('select').simulate('change', e);
  expect(sortByAmount).toHaveBeenCalled();
});

test('should handle date changes', () => {
  const args = {
    startDate: 1,
    endDate: 2
  };
  wrapper.find('withStyles(DateRangePicker)').prop('onDatesChange')(args);
  expect(setStartDate).toHaveBeenLastCalledWith(args.startDate);
  expect(setEndDate).toHaveBeenLastCalledWith(args.endDate);
});

test('should handle date focus changes', () => {
  const calendarFocussed = 'endDate';
  wrapper.find('withStyles(DateRangePicker)').prop('onFocusChange')(calendarFocussed);
  expect(wrapper.state('calendarFocussed')).toEqual(calendarFocussed);
});
