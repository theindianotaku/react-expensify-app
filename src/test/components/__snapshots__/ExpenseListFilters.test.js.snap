// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListFilter properly 1`] = `
<div
  className="content-container"
>
  <div
    className="input-group"
  >
    <div
      className="input-group__item"
    >
      <input
        className="text-input"
        onChange={[Function]}
        placeholder="Search expenses"
        type="text"
        value=""
      />
    </div>
    <div
      className="input-group__item"
    >
      <select
        className="select"
        onChange={[Function]}
      >
        <option
          value="date"
        >
          Date
        </option>
        <option
          value="amount"
        >
          Amount
        </option>
      </select>
    </div>
    <div
      className="input-group__item"
    >
      <withStyles(DateRangePicker)
        anchorDirection="left"
        customArrowIcon={null}
        customCloseIcon={null}
        customInputIcon={null}
        daySize={39}
        disabled={false}
        displayFormat={[Function]}
        enableOutsideDays={false}
        endDate={null}
        endDateId="endDate"
        endDatePlaceholderText="End Date"
        firstDayOfWeek={null}
        focusedInput={null}
        hideKeyboardShortcutsPanel={false}
        horizontalMargin={0}
        initialVisibleMonth={null}
        inputIconPosition="before"
        isDayBlocked={[Function]}
        isDayHighlighted={[Function]}
        isOutsideRange={[Function]}
        isRTL={false}
        keepOpenOnDateSelect={false}
        minimumNights={1}
        monthFormat="MMMM YYYY"
        navNext={null}
        navPrev={null}
        numberOfMonths={1}
        onClose={[Function]}
        onDatesChange={[Function]}
        onFocusChange={[Function]}
        onNextMonthClick={[Function]}
        onPrevMonthClick={[Function]}
        openDirection="down"
        orientation="horizontal"
        phrases={
          Object {
            "calendarLabel": "Calendar",
            "chooseAvailableEndDate": [Function],
            "chooseAvailableStartDate": [Function],
            "clearDates": "Clear Dates",
            "closeDatePicker": "Close",
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "focusStartDate": "Interact with the calendar and add the check-in date for your trip.",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month",
            "jumpToPrevMonth": "Move backward to switch to the previous month",
            "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
            select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
            "moveFocusByOneMonth": "Switch months",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
            "openThisPanel": "Open this panel",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field",
            "selectFocusedDate": "Select the date in focus",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
            "upArrowDownArrow": "up and down arrow keys",
          }
        }
        readOnly={false}
        renderCalendarInfo={null}
        renderDay={null}
        renderMonth={null}
        reopenPickerOnClearDates={false}
        required={false}
        screenReaderInputMessage=""
        showClearDates={true}
        showDefaultInputIcon={false}
        startDate={null}
        startDateId="startDate"
        startDatePlaceholderText="Start Date"
        verticalHeight={null}
        weekDayFormat="dd"
        withFullScreenPortal={false}
        withPortal={false}
      />
    </div>
  </div>
</div>
`;

exports[`should render ExpenseListFilter with alt data properly 1`] = `
<div
  className="content-container"
>
  <div
    className="input-group"
  >
    <div
      className="input-group__item"
    >
      <input
        className="text-input"
        onChange={[Function]}
        placeholder="Search expenses"
        type="text"
        value="bill"
      />
    </div>
    <div
      className="input-group__item"
    >
      <select
        className="select"
        onChange={[Function]}
      >
        <option
          value="date"
        >
          Date
        </option>
        <option
          value="amount"
        >
          Amount
        </option>
      </select>
    </div>
    <div
      className="input-group__item"
    >
      <withStyles(DateRangePicker)
        anchorDirection="left"
        customArrowIcon={null}
        customCloseIcon={null}
        customInputIcon={null}
        daySize={39}
        disabled={false}
        displayFormat={[Function]}
        enableOutsideDays={false}
        endDate={"1970-01-06T00:00:00.000Z"}
        endDateId="endDate"
        endDatePlaceholderText="End Date"
        firstDayOfWeek={null}
        focusedInput={null}
        hideKeyboardShortcutsPanel={false}
        horizontalMargin={0}
        initialVisibleMonth={null}
        inputIconPosition="before"
        isDayBlocked={[Function]}
        isDayHighlighted={[Function]}
        isOutsideRange={[Function]}
        isRTL={false}
        keepOpenOnDateSelect={false}
        minimumNights={1}
        monthFormat="MMMM YYYY"
        navNext={null}
        navPrev={null}
        numberOfMonths={1}
        onClose={[Function]}
        onDatesChange={[Function]}
        onFocusChange={[Function]}
        onNextMonthClick={[Function]}
        onPrevMonthClick={[Function]}
        openDirection="down"
        orientation="horizontal"
        phrases={
          Object {
            "calendarLabel": "Calendar",
            "chooseAvailableEndDate": [Function],
            "chooseAvailableStartDate": [Function],
            "clearDates": "Clear Dates",
            "closeDatePicker": "Close",
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "focusStartDate": "Interact with the calendar and add the check-in date for your trip.",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month",
            "jumpToPrevMonth": "Move backward to switch to the previous month",
            "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
            select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
            "moveFocusByOneMonth": "Switch months",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
            "openThisPanel": "Open this panel",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field",
            "selectFocusedDate": "Select the date in focus",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
            "upArrowDownArrow": "up and down arrow keys",
          }
        }
        readOnly={false}
        renderCalendarInfo={null}
        renderDay={null}
        renderMonth={null}
        reopenPickerOnClearDates={false}
        required={false}
        screenReaderInputMessage=""
        showClearDates={true}
        showDefaultInputIcon={false}
        startDate={"1970-01-01T00:00:00.000Z"}
        startDateId="startDate"
        startDatePlaceholderText="Start Date"
        verticalHeight={null}
        weekDayFormat="dd"
        withFullScreenPortal={false}
        withPortal={false}
      />
    </div>
  </div>
</div>
`;
